python/hail/history.py

```diff
@@ -0,0 +1,193 @@
+    return result
+
+
+def write_history(path_arg_name, is_dir=False):
+    def _write(f, obj, *args, **kwargs):
+        parsed_args = parse_args(f, args, kwargs)
+        argnames = inspect.getcallargs(f, obj, *args, **kwargs)
+        result = f(obj, *args, **kwargs)
+
+        output_path = argnames[path_arg_name]
+        if is_dir:
+            output_path = output_path + "/history.txt"
+        else:
+            output_path += ".history.txt"
+
+        (obj._history
+         .add_method(f.__name__, parsed_args)
+         .write(output_path))
+
+        return result
+    return decorator(_write)
+
+
+def format_args(arg, statements):
+    if isinstance(arg, list):
+        return [format_args(a, statements) for a in arg]
+    elif isinstance(arg, tuple):
+        return tuple([format_args(a, statements) for a in arg])
+    elif isinstance(arg, dict):
+        return {format_args(k, statements): format_args(v, statements) for k, v in arg.iteritems()}
+    else:
+        if isinstance(arg, HistoryMixin):
+            h = arg._history
+            statements += remove_dup_statements(statements, h.statements)
+            return h
+        else:
+            return arg
+
+
+def remove_dup_statements(statements1, statements2):
+    return [s2 for s1, s2 in itertools.izip_longest(statements1, statements2) if s2 and s1 != s2]
+
+
+class History(object):
+    def __init__(self, expr="", statements=[]):
+        self.expr = expr
+        self.statements = statements
+
+    def set_varid(self, id):
+        return History(id, self.statements + ['{} = ({})'.format(id, self.expr)])
+
+    def add_method(self, f_name, kwargs, index=None, key_name=None):
+        statements = self.statements[:]
+        f_args = ["{}={}".format(k, repr(format_args(v, statements))) for k, v in kwargs.iteritems()]
+        expr = "{expr}\n.{f_name}({f_args})".format(expr=self.expr, f_name=f_name, f_args=", ".join(f_args))
+        if index:
+            expr += "[{}]".format(index)
+        elif key_name:
+            expr += "[{}]".format(repr(key_name))
+        return History(expr, statements)
+
+    @staticmethod
+    def from_classmethod(cls_name, f_name, kwargs):
+        statements = []
+        f_args = ["{}={}".format(k, repr(format_args(v, statements))) for k, v in kwargs.iteritems()]
+        expr = "{cls_name}.{f_name}({args})".format(cls_name=cls_name, f_name=f_name, args=", ".join(f_args))
+        return History(expr, statements)
+
+    @staticmethod
+    def from_init(cls_name, kwargs):
+        statements = []
+        f_args = ["{}={}".format(k, repr(format_args(v, statements))) for k, v in kwargs.iteritems()]
+        expr = "{cls_name}({args})".format(cls_name=cls_name, args=", ".join(f_args))
+        return History(expr, statements)
+
+    def write(self, file):
+        with hadoop_write(file) as f:
+            f.write(str(self) + "\n")
+
+    def __repr__(self):
+        return self.expr
+
+    def __str__(self):
```

Comment:
I don't quite like that the __str__ of history is the script. Can we move this to a new method like `formatted`? And can we add the `from hail import *` at the top?